### PR TITLE
svnauthzsub

### PR DESCRIPTION
**Pull request by omidontop**
Resolves: https://pds-svn-gbg.pdsvision.net/trac/SubversionCMS/ticket/1673#comment:12
Related: https://github.com/simonsoft/chef/pull/411

### CODE REVIEW

**Comment by takesson** [comment]
Nice, is there a command for running python unit tests and getting a Jenkins-compatible report? I think I have seen junit compatible report in Python.

**Comment by takesson** [comment]
There was a late change / realisation in Apache config that requires repo root Location to trim the trailing slash. E.g.:
```
<Location /svn/repo1 >
```

**Comment by omidontop** [comment]
> Nice, is there a command for running python unit tests and getting a Jenkins-compatible report? I think I have seen junit compatible report in Python.

Yep, this one:

`nose2 -v --plugin nose2.plugins.junitxml --junit-xml svnauthzsubtest.SvnAuthzSubTestCase`

**Comment by omidontop** [comment]
> There was a late change / realisation in Apache config that requires repo root Location to trim the trailing slash. E.g.:
> 
> ```
> <Location /svn/repo1 >
> ```

Done.

File: svnauthzsub.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+import configparser
+import svnpubsub.logger
+from io import StringIO
+from textwrap import indent
+from urllib.request import urlopen
+from svnpubsub.client import Commit
+from svnpubsub.daemon import Daemon, DaemonTask
+from svnpubsub.bgworker import BackgroundJob
+
+MC_PORT = 2069
```
Comment (takesson) [inline svnauthzsub.py:43]: I think we should keep just "PORT" for symmetry with the other subscribing services.

File: svnauthzsub.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+            logging.error("Failed to retrieve the access.accs file: %s", str(e))
+        return None
+
+    def validate(self) -> bool:
+        return True
+
+    def run(self):
+        global OUTPUT_DIR
+        access_accs = self.retrieve_access_accs()
+        print(access_accs)
```
Comment (takesson) [inline svnauthzsub.py:147]: Debug statement?

Comment (omidontop) [reply svnauthzsub.py]: Yes, removed.

File: svnauthzsub.py (hunk 1 above)
Comment (omidontop) [reply svnauthzsub.py]: Ok

**Comment by takesson** [comment]
This odd (but valid in svnauthz) statement generates invalid conf:
```
[/qa/readonly]
* = r
```
All statements starting with * (all users) should be ignored, logged if non-empty on the right hand side.
There is no issue for us as long as we don't support inheritance.

**Comment by takesson** [comment]
Have we covered (test coverage preferably) the case when a section ONLY has 
```
* =
```
I think we need to handle that as a special case and produce:
```
Require all denied
```

File: svnauthzsub.py [hunk 3]
```diff
@@ -48,13 +50,35 @@
...
+    result = True
+    config_parser = ConfigParser()
+    config_parser.read_string(os.linesep.join(access_accs) if isinstance(access_accs, list) else access_accs)
+    paths = [section for section in config_parser.sections() if '/' in section]
+    for path in paths:
+        section = config_parser[path]
+        roles = [role for role in section if role.startswith('@')]
+        # Path may not contain extended characters as Apache doesn't support UTF8 locations
+        # Accept alphanumeric characters plus the following: _./ -
+        if not re.match("^([A-Za-z0-9_./ -]*)$", path, re.RegexFlag.MULTILINE):
```
[the remark this answers is not in the document]
Comment (takesson) [reply svnauthzsub.py:70]: But not in Python.

File: svnauthzsub.py [hunk 4]
```diff
@@ -197,24 +223,34 @@ def main():
...
-                exit(0)
-        except Exception as e:
-            logging.error("%s", str(e))
-            exit(1)
-    elif not args.output_dir:
-        parser.error('OUTPUT_DIR must be provided')
+        # The input has been provided through stdin
+        elif select.select([sys.stdin, ], [], [], 0.0)[0]:
+            access_accs = sys.stdin.read()
+        # No explicit input has been provided, the OUTPUT_DIR becomes mandatory
```
[the remark this answers is not in the document]
Comment (omidontop) [reply svnauthzsub.py:236]: Not exactly, there is another option and that is to provide the input through the --input file and the aim might be to write the output to an --output file or to the stdout.

File: svnauthzsub.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (omidontop) [reply svnauthzsub.py:70]: Wanted to be safe. OK.

File: svnauthzsub.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (takesson) [reply svnauthzsub.py:236]: Ok... I really struggle to wrap my head around a switch conditional based on stdin data existence.

[the remark this answers is not in the document]
Comment (takesson) [reply svnauthzsub.py:236]: Typically a command determined that it should read from stdin based on switches and then hangs until data arrives.

[the remark this answers is not in the document]
Comment (omidontop) [reply svnauthzsub.py]: Done.

File: svnauthzsub.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (omidontop) [reply svnauthzsub.py]: Removed.

**Pull request merged**